client/src/api.ts

```diff
@@ -1,12 +1,59 @@
 import useSWR, { responseInterface } from 'swr';
+import jwtDecode from 'jwt-decode';
 
-import { useStore } from './store';
+import { useStore, Action } from './store';
+
+interface DecodedToken {
+  token_type: string;
+  exp: number;
+  user_id: number;
+}
+
+/**
+ * A react hook to get a valid access token. It will refresh the token if needed.
+ * Returns the token if it's possible to get it, otherwise returns undefined.
+ */
+export async function getAccessToken(
+  tokens: { accessToken: string | null; refreshToken: string | null },
+  base: string,
+  dispatch: React.Dispatch<Action>
+): Promise<string | undefined> {
+  const { accessToken, refreshToken } = tokens;
+  // We can't get a token if we don't have a refresh token or an access token
+  if (!accessToken && !refreshToken) return undefined;
+  // Return the token if the one we have is not expired
+  const decoded: DecodedToken = jwtDecode(accessToken as string);
+  if (decoded.exp > Date.now() / 1000) return accessToken as string;
+  // If the token is expired, refresh
+  const newToken = await apiReq(base, 'token/refresh', {
+    // To refresh, we send a POST request with the refresh token
+    method: 'POST',
+    body: JSON.stringify({ refresh: refreshToken }),
+    headers: { 'Content-Type': 'application/json' },
+  })
+    .then(({ access: fetchedToken }: { access: string }) => {
+      dispatch({
+        type: 'login',
+        tokens: { access: fetchedToken, refresh: refreshToken as string },
+      });
+      return fetchedToken;
+    })
+    .catch((e) => {
+      // If we get an error, return undefined and log the user out
+      console.error('Error during token refresh:');
+      console.log(e);
+      dispatch({ type: 'logout' });
+      return undefined;
```

Comment:
If token authentication fails, this code will report the error that occured to the console, and then automatically kick the user out to the "login" screen (by dispatching the logout action). I think this is the most useful behavior: the error can be inspected and the user has to log in again to rectify the situation. If the error was thrown instead of reported with `console.error`, the app would crash (and in development, unmount) whenever token refresh could not be completed. What would be better about that behavior?

Reply:
Gotcha. It seems like there are many of these functions returning undefined on failure. Maybe we should implement an [Option/Maybe type](https://softwareengineering.stackexchange.com/questions/385632/is-there-a-convention-for-the-optional-maybe-monad-in-typescript) or at least use `null` instead of `undefined`.